Aap 2007 AAP 2.1 on OCP Pre-Flight Check Documentation (#332)
RedHatInsights/red-hat-ansible-automation-platform-documentation: downstream/modules/platform/con_storage-options-for-operator-installation-on-ocp.adoc

[id="con_storage-options-for-operator-installation-on-ocp_{context}"]
= Storage options for {OperatorPlatform} installation on {OCP}

> If your installation scenario includes {HubName}, you must ensure that your storage option for {OperatorPlatform} has been set to `ReadWriteMany`.

Reviewer: This is true if/when the customer is using "file" based storage. If they selected S3 or Azure than this access mode doesn't mean anything.

Author: Thanks for the feedback here. Was unsure about the technical accuracy. Will propose a new statement.

> {HubNameStart} requires `ReadWriteMany` storage for operation or Azure Blob storage or an S3-compliant storage because multiple pods responsible for upload and download of content (for example, collections) need access to the same content. Otherwise, multiple pods are not guaranteed to have access to the same permanent virtual circuit (PVC)  unless it is `ReadWriteMany`.

Reviewer: PVC here is "persistent volume claim" not "permanent virtual circuit".

Author: Thank you for the correction.

In addition, OpenShift Data Foundation provides either or both a ReadWriteMany or s3-compliant implementation. Also, an NFS storage configuration can be setup to support `ReadWriteMany` and is often the simplest approach. This however, introduces the NFS server as a potential single point of failure.

[role="_additional-resources"]
.Additional resources

> * IBM's link:https://www.ibm.com/support/pages/how-do-i-create-storage-class-nfs-dynamic-storage-provisioning-openshift-environment[How do I create a storage class for NFS dynamic storage provisioning in an OpenShift environment?]

Reviewer: We should probably promote a Red Hat link:
https://docs.openshift.com/container-platform/4.10/storage/persistent_storage/persistent-storage-nfs.html

I think the IBM link is a better walk-thru but I don't know if we typically reference docs outside of Red Hat for how to configure OpenShift.

> [id="con_storage-options-for-operator-installation-on-ocp_{context}"]

Reviewer 2: I think we should remove the prefix from the ID for SEO purposes. The ID forms part of the URL and when that includes `con_` etc it is less meaningful. Basically it's just noise to the search engine.

> = Storage options for {OperatorPlatform} installation on {OCP}

Reviewer 2: s/options/requirements 

I don't think RWX storage is optional for the hub.

Author: I admit I have a technological gap here with this product. Are you referring to the title for this? The SME, Chris Hambridge, reviewed this a couple of weeks ago and did not mention the title being erroneous. That said, it doesn't mean there might be a better way of describing the content. I'll have to circle back to check on this, but he didn't mention any issues with the title. Going with "stet" with this one for now.

> If you are using file-based storage and your installation scenario includes {HubName}, you must ensure that your storage option for {OperatorPlatform} has been set to `ReadWriteMany`.

Reviewer 3: ```suggestion
If you are using file-based storage and your installation scenario includes {HubName}, ensure that the storage option for {OperatorPlatform} is set to `ReadWriteMany`.
```

Reviewer 3: Out of scope for this PR, I know, but I'm curious whether there's a default setting for the storage option.

Reviewer 3: @jself-sudoku It's not essential, but it might be a good idea in this case, because both option names begin with `ReadWrite`.

Author: Thanks, @ariordan-redhat. I have updated the intro sentence to show the default setting.

> If you are using file-based storage and your installation scenario includes {HubName}, ensure that you change the `ReadWriteOnce` default storage option for {OperatorPlatform} to `ReadWriteMany`.

Reviewer 2: ```suggestion
If your installation includes {HubName}, you must provision a persistent volume claim with `ReadWriteOnce` access.
```

Reviewer 2: something along those lines imho

Author: Actually, the goal with this is to provision with `ReadWriteMany` access mode. However, I do like the concise way of stating this, which is similar to my original statement. I updated the statement because of a peer review comment that I received previously that it might be good to mention the default option so that the customer can distinguish the two, I guess. I'll return to the original statement that keeps things simplified.

> {HubNameStart} requires `ReadWriteMany` storage for operation or Azure Blob storage or an S3-compliant storage because multiple pods responsible for upload and download of content (for example, collections) need access to the same content. Otherwise, multiple pods are not guaranteed to have access to the same persistent volume claim (PVC)  unless it is `ReadWriteMany`.

Reviewer 3: ```suggestion
{HubNameStart} requires `ReadWriteMany` storage or Azure Blob storage or an S3-compliant storage for operation because multiple pods responsible for upload and download of content (for example, collections) need access to the same content. Multiple pods are guaranteed to have access to the same persistent volume claim (PVC) only if `ReadWriteMany` storage is enabled.
```

> {HubNameStart} requires `ReadWriteMany` storage, Azure Blob storage, or an s3-compliant storage for operation because multiple pods responsible for upload and download of content (for example, collections) need access to the same content. Multiple pods can have access to the same persistent volume claim (PVC) as long as you select the `ReadWriteMany` storage option.

Reviewer 2: ```suggestion
{HubNameStart} requires `ReadWriteMany` storage, Azure Blob storage, or an s3-compliant storage for operation so that multiple pods can access shared content, such as collections.
```

Reviewer 2: I don't think the 2nd sentence is necessary for minimalism. You're basically explaining RWX access, which is covered in k8s and OCP docs.

Author: Well, you have just cleared up the thing I was most confused about. My technological gap had me unsure about it, but since you're are knowledgeable about RWX, it clears things up. There were too many "or" conjunctions, which was muddying the waters, so to speak. Updating accordingly.

> In addition, OpenShift Data Foundation provides a `ReadWriteMany` or s3-compliant implementation. Also, an NFS storage configuration can be setup to support `ReadWriteMany` and is often the simplest approach. This, however, introduces the NFS server as a potential single point of failure.

Reviewer 3: Is S3 supposed to be uppercase or lowercase?

Reviewer 3: Do you think it would be useful to have an attribute for OpenShift Data Foundation?

Author: I actually considered that. I have added two entries to the attributes file for this: 

:ODF: Red Hat OpenShift Data Foundation
:ODFShort: OpenShift Data Foundation

> In addition, OpenShift Data Foundation provides a `ReadWriteMany` or S3-compliant implementation. Also, an NFS storage configuration can be setup to support `ReadWriteMany` and is often the simplest approach. This, however, introduces the NFS server as a potential single point of failure.

Reviewer 2: Why do we mention Data Foundation here? Should we use the variable instead `{ODFShort}`?

Reviewer 2: "simplest approach" compared to what? I'm not sure what this is telling the reader to do with storage. Can you use ODF as the underlying storage mechanism?

There's also passive voice in the second sentence "can be setup". That should be a phrasal verb rather than noun form as well. s/setup/set up

Author: I just added OpenShift Data Foundation to the attributes file yesterday, per @ariordan-redhat 's request. So this is now updated.

[role="_additional-resources"]
.Additional resources

> * link:https://docs.openshift.com/container-platform/4.10/storage/persistent_storage/persistent-storage-nfs.html[Persistent storage using NFS]

Reviewer 3: ```suggestion
* link:https://docs.openshift.com/container-platform/{OCPLatest}/storage/persistent_storage/persistent-storage-nfs.html[Persistent storage using NFS] in the {OCPShort} _Storage_ guide
```

Just a suggestion: indicate that the link to the instructions for NFS are in the OpenShift docs
